debianutils-2.6.1/run-parts.8
.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.\" Build-from-directory and this manpage are Copyright 1994 by Ian Jackson.
.\" Changes to this manpage are Copyright 1996 by Jeff Noxon.
.\" More
.\"
.\" This is free software; see the GNU General Public Licence version 2
.\" or later for copying conditions.  There is NO warranty.
.TH RUN\-PARTS 8 "12 November 2003" "Debian"
.SH NAME
run\-parts \- run scripts or programs in a directory
.SH SYNOPSIS
.B run\-parts
[\-\-test] [\-\-verbose] [\-\-report] [\-\-lsbsysinit] [\-\-umask=umask]
[\-\-arg=argument] [\-\-exit\-on\-error] [\-\-help] [\-\-version]
[\-\-list] [\-\-reverse] [\-\-] directory
.SH DESCRIPTION
.PP
.B run\-parts
runs all the executable files found in directory
.IR directory .
Other files and directories are silently ignored.

If the \-\-lsbsysinit option is not given then the names must consist
entirely of upper and lower case letters, digits, underscores, and
hyphens.                           
                                                                                
If the \-\-lsbsysinit option is given, then the names must not end
in .dpkg\-old  or .dpkg\-dist and must belong to one or more of the
following namespaces: the LANANA-assigned namespace (^[a\-z0\-9]+$);
the LSB hierarchical and reserved namespaces (^_?([a\-z0\-9_.]+\-)+[a\-z0\-9]+$);
and the Debian cron script namespace (^[a\-z0\-9][a\-z0\-9\-]*$).      

Files are run in the lexical sort order of their names unless the
\-\-reverse option is given, in which case they are run in the
opposite order.

.SS OPTIONS
.TP
.B \-\-test
print the names of the scripts which would be run, but don't actually run
them.
.TP
.B \-\-list
print the names of the all matching files (not limited to executables),
but don't actually run them. This option cannot be used with --test.
.TP
.B \-\-verbose
print the name of each script to stderr before running.
.TP
.B \-\-report
similar to
.BR \-\-verbose ,
but only prints the name of scripts which produce output.  The script's name is printed to whichever of stdout or stderr the script first produces output on.
.TP
.B \-\-reverse
reverse the scripts' execution order.
.B
.TP
.B \-\-exit\-on\-error
exit as soon as a script returns with a non-zero exit code.
.TP
.B \-\-lsbsysinit
Use LSB namespaces instead of classical behavior.
.TP
.BI \-\-umask= umask
Sets the umask to
.I umask
before running the scripts.
.I umask
should be specified in octal.  By default the umask is set to 022.
.TP
.BI \-\-arg= argument
Pass
.I argument
to the scripts.  Use
.B --arg
once for each argument you want passed.
.TP
.B "\-\-"
Specifies that this is the end of the options.  Any filename after
.B "\-\-"
will be not be interpreted as an option even if it starts with a
hyphen.
.TP
.B \-\-help
Display usage information and exit.
.TP
.B \-\-version
Display version and copyright and exit.

.SH COPYRIGHT
Copyright (C) 1994 Ian Jackson.
Copyright (C) 1996 Jeff Noxon.
Copyright (C) 1996,1997,1998 Guy Maor

.B run\-parts
is free software; see the GNU General Public Licence version 2 or
later for copying conditions.  There is
.I no
warranty.


proftpd/NEWS
$Id: NEWS,v 1.594 2004/09/04 18:50:03 castaglia Exp $

1.2.10 - Released 04-Sep-2004
--------------------------------
- Bug 2440 - Unable to use PAM authentication properly.  Use a "*" after
  the module name in an AuthOrder directive to indicate that an auth
  module is authoritative.
- Bug 2441 - AIX5 portability bug with mod_auth_unix, mod_auth_file.
- Bug 2442 - Segfault in FreeBSD PAM library with long login names.
- Bug 2445 - AuthUserFile in <Global> context overrides <VirtualHost> setting.
- Bug 2444 - Use of sendfile() does not interoperate well with RFC2228
  security mechanisms.  Using sendfile(2) to send data bypasses the handling
  of the data by RFC2228 security mechanisms (such as those provided by
  mod_tls).  So if security mechanisms are detected, do not use sendfile().
- Scrub the ScoreboardFile for stale sessions in inetd mode.
- Bug 2427 - proftpd gets a memory fault when run from ssh batch mode.

1.2.10rc3 - Released 13-Jul-2004
--------------------------------
- Fixed typo that prevented 1.2.10rc2 from compiling.

1.2.10rc2 - Released 13-Jul-2004
---------------------------------
- Bug 2396 - NLST command doesn't understand options.  This was caused by
  the solution for Bug 2322.  However, it is not a popular solution, so
  NLST will once again handle options, but only the relevant options.
- Bug 2034 - Add support for a "graceful shutdown" signal. See
  contrib/mod_ctrls_admin.html#shutdown for details.
- Bug 2400 - <Class> search order is wrong.  The documentation correctly
  stated that <Class> sections are matched in order of definition, but the
  code has the match order in the reverse order of definition.
- Bug 2401 - MaxClientsPerClass only checks first directive in config file.
- Bug 2399 - Rename start/stop control actions to up/down.
- Bug 2082 - Add mod_rewrite "replaceall" builtin function.  See the
  RewriteMap documentation for more details.
- Bug 2403 - Sending SIGHUP to proftpd stops it when using Classes.  The fix
  for Bug #2400 could result in an infinite loop during a SIGHUP.
- Bug 2405 - "LIST *" should not list dotfiles.
- Bug 2366 - Add support for -h list option.
- Bug 2332 - SO_OOBINLINE error after upgrading proftpd from 1.2.6 to 1.2.9.
  This is due mostly to a change in the logging; a check for error values
  and logging of them was added.  The setting of this particular socket option
  has been moved earlier in the session, as it was found that short-lived
  TCP connections, as from monitoring systems, would cause this error.
- Bug 2407 - mod_auth_file does not allow for proper cascading of "end" and
  "set" auth requests.
- Bug 2410 - CreateHome always copies skel directory.
- Bug 2336 - Use of /dev/log on Solaris leads to kernel memory leak.
  ProFTPD's use of the /dev/log device on Solaris was tickling a Solaris
  kernel bug that caused the Solaris kernel to leak memory.
- Added a TimeoutLinger directive to complement the --enable-timeout-linger
  configure option.
- Bug 2125 - -vv command line switch should list versions of modules.
- Bug 2420 - Name field is not escaped before querying database.
  The mod_quotatab module was not properly escaping the name string it
  used when looking up records from SQL databases.
- Bug 2424 - SQLDefaultHomedir overrides column value.
- Bug 2411 - Caching effects cause RNTO to fail if AllowOverwrite is off and
  target path does not exist.
- Bug 2422 - %v not working in SQLNamedQuery.
- Bug 2418 - chmod returns 550 with filename containing multiple spaces.
- Bug 2431 - mod_sql does not use UID/GID properly in cache lookups.
- Bug 2303 - Problem evaluating multiple <Class> rules.
- Bug 2419 - Ability to disable TLSRequired on per-user basis (e.g. for
  anonymous logins).
- Bug 2438 - Display variable %z not expanded properly.
- Bug 2439 - <Limit CWD> doesn't work.

1.2.10rc1 - Released 28-Apr-2004
---------------------------------
- Bug 2135 - Add ability to handle passphrase-protected server keys.  mod_tls
  can now properly prompt for passphrases for protected server certificate
  keys when the daemon is starting up.
- Bug 2086 - Add limits for PORT, PASV.  This means that now one can use
  <Limit> to place access controls on the PORT and PASV commands.  This
  applies to the EPRT and EPSV commands as well.
- Bug 2174 - mod_auth_unix should not act authoritatively.  This was causing
  problems when using mod_auth_unix.c and the AuthOrder configuration
  directive.
- Bug 2098 - Added SetEnv and UnsetEnv configuration directives.
- Bug 2271 - Improper autoconf check for getaddrinfo() on Tru64 UNIX 5.1.
  The getaddrinfo symbol is a macro, not a function, on that platform.
- Bug 2255 - RADIUS Service-Type should reflect attribute expectations.
- Added Event API.
- Bug 2272 - Address/port collision check needs to handle DefaultAddress.
- Bug 2072 - Add Controls API.
  This API includes a new program, ftpdctl, that is used to communicate
  directly with the proftpd daemon via a new core module, mod_ctrls.  For
  this new functionality to be used, proftpd must be configured using the
  added --enable-ctrls option.
- Bug 2015 - Add AND, OR keywords to Allow/DenyUser directives.
  The AllowUser, DenyUser, AllowGroup, and DenyGroup directives now take
  an optional keyword that indicates what type of expression they are:
  AND, OR, or regex.  By default, AllowUser and DenyUser are OR expressions,
  and AllowGroup and DenyGroup are AND expressions.  For example:

    AllowUser regex ^ftp
    DenyUser AND dave,bob
    AllowGroup OR web,doc

  These demonstrate that the optional keyword modifier must be the first
  parameter in the configuration directive.
- Bug 2046 - Change RFNR and RNTO logging class to WRITE.  This means that
  ExtendedLogs that use the WRITE logging class will now include the
  RNFR and RNTO commands.
- Mac OS X 10.3 portability fixes.
- Bug 2274 - Default server only binds to one IP address of host if
  the --enable-ipv6 configure option is used.  ProFTPD will now properly

debianutils-2.6.1/run-parts.c
/* run-parts: run a bunch of scripts in a directory
 *
 * Debian run-parts program
 * Copyright (C) 1996 Jeff Noxon <jeff@router.patch.net>,
 * Copyright (C) 1996-1999 Guy Maor <maor@debian.org>
 *
 * This is free software; see the GNU General Public License version 2
 * or later for copying conditions.  There is NO warranty.
 *
 * Based on run-parts.pl version 0.2, Copyright (C) 1994 Ian Jackson.
 *
 */

#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <dirent.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <signal.h>
#include <sys/time.h>
#include <regex.h>

int test_mode = 0;
int list_mode = 0;
int verbose_mode = 0;
int report_mode = 0;
int reverse_mode = 0;
int exitstatus = 0;
int lsbsysinit_mode = 0;
int exit_on_error_mode = 0;

int argcount = 0, argsize = 0;
char **args = 0;

void error(char *format, ...)
{
  va_list ap;

  fprintf(stderr, "run-parts: ");

  va_start(ap, format);
  vfprintf(stderr, format, ap);
  va_end(ap);

  fprintf(stderr, "\n");
}


void version()
{
  fprintf(stderr, "Debian run-parts program, version " PACKAGE_VERSION
	  "\nCopyright (C) 1994 Ian Jackson, Copyright (C) 1996 Jeff Noxon.\n"
	  "Copyright (C) 1996,1997,1998,1999 Guy Maor\n"
	  "This is free software; see the GNU General Public License version 2\n"
	  "or later for copying conditions.  There is NO warranty.\n");
  exit(0);
}


void usage()
{
  fprintf(stderr, "Usage: run-parts [OPTION]... DIRECTORY\n"
	  "  --test          print script names which would run, but don't run them.\n"
	  "  --list          print names of all valid files (can not be used with --test)\n"
	  "  --verbose       print script names before running them.\n"
	  "  --report        print script names if they produce output.\n"
	  "  --reverse       reverse execution order of scripts.\n"
	  "  --exit-on-error exit as soon as a script returns with a non-zero exit code.\n"
	  "  --lsbsysinit    validate filenames based on LSB sysinit specs.\n"
	  "  --umask=UMASK   sets umask to UMASK (octal), default is 022.\n"
	  "  --arg=ARGUMENT  pass ARGUMENT to scripts, use once for each argument.\n"
	  "  --version       output version information and exit.\n"
	  "  --help          display this help and exit.\n");
  exit(0);
}


/* The octal conversion in libc is not foolproof; it will take the 8 and 9
 * digits under some circumstances.  We'll just have to live with it.
 */
void set_umask()
{
  int mask, result;

  result = sscanf(optarg, "%o", &mask);
  if ((result != 1) || (mask > 07777) || (mask < 0)) {
    error("bad umask value");
    exit(1);
  }

  umask(mask);
}

/* Add an argument to the commands that we will call.  Called once for
   every argument. */
void add_argument(char *newarg)
{
  if (argcount + 1 >= argsize) {
    argsize = argsize ? argsize * 2 : 4;
    args = realloc(args, argsize * (sizeof(char *)));
    if (!args) {
      error("failed to reallocate memory for arguments: %s", strerror(errno));
      exit(1);
    }
  }
  args[argcount++] = newarg;
  args[argcount] = 0;
}

/* True or false? Is this a valid filename? */
int valid_name(const struct dirent *d)
{
  char *c = d->d_name;
  regex_t hierre, tradre, excsre, classicalre;

  /* The regcomps should be moved to program init */

  if (lsbsysinit_mode) {

    if (regcomp
	(&hierre, "^_?([a-z0-9_.]+-)+[a-z0-9]+$", REG_EXTENDED | REG_NOSUB)
	|| regcomp(&excsre, "^[a-z0-9-].*dpkg-(old|dist)$",
		   REG_EXTENDED | REG_NOSUB)
	|| regcomp(&tradre, "^[a-z0-9][a-z0-9-]*$", REG_NOSUB)) {
      error("regex failure");
      exit(1);
    }

    if (!regexec(&hierre, c, 0, NULL, 0))
      return regexec(&excsre, c, 0, NULL, 0);

    return !regexec(&tradre, c, 0, NULL, 0);
  }
  else {
    if (regcomp(&classicalre, "^[a-zA-Z0-9_-]+$", REG_EXTENDED | REG_NOSUB)) {
      error("regex failure");
      exit(1);
    }
    return !regexec(&classicalre, c, 0, NULL, 0);
  }

}


void set_fl(int fd, int flags)
{
  int val;
  if ((val = fcntl(fd, F_GETFL, 0)) < 0) {
    error("fcntl F_GETFL: %s", strerror(errno));
    exit(1);
  }
  val |= flags;
  if (fcntl(fd, F_SETFL, val) < 0) {
    error("fcntl F_SETFL: %s", strerror(errno));
    exit(1);
  }
}


/* We have to abort the select() call in run_part when we receive a
 * SIGCHLD signal. We can't simply ignore it, so we do nothing.
 */

/* itz Tue Nov 26 13:23:27 PST 2002 why "We have to abort?"
   just leave the default and reap child at the end
 */
void catch_sigchld(int sig)
{
}


/* Execute a file */
void run_part(char *progname)
{
  int result;
  int pid;
  int pout[2], perr[2];

  if (report_mode && (pipe(pout) || pipe(perr))) {
    error("pipe: %s", strerror(errno));
    exit(1);
  }
  if ((pid = fork()) < 0) {
    error("failed to fork: %s", strerror(errno));
    exit(1);
  }
  else if (!pid) {
    if (report_mode) {
      if (dup2(pout[1], STDOUT_FILENO) == -1 ||
	  dup2(perr[1], STDERR_FILENO) == -1) {
	error("dup2: %s", strerror(errno));
	exit(1);
      }
      close(pout[0]);
      close(perr[0]);
      close(pout[1]);
      close(perr[1]);
    }
    args[0] = progname;
    execv(progname, args);
    error("failed to exec %s: %s", progname, strerror(errno));
    exit(1);
  }

  if (report_mode) {
    fd_set set;
    int max, r, printflag;
    ssize_t c;
    char buf[4096];

    close(pout[1]);
    close(perr[1]);
    /* itz Tue Nov 26 13:26:09 PST 2002 why is this necessary? */
    /*     set_fl(pout[0], O_NONBLOCK); */
    /*     set_fl(perr[0], O_NONBLOCK); */
    max = pout[0] > perr[0] ? pout[0] + 1 : perr[0] + 1;
    printflag = 0;

    while (pout[0] >= 0 || perr[0] >= 0) {

      FD_ZERO(&set);
      if (pout[0] >= 0)
	FD_SET(pout[0], &set);
      if (perr[0] >= 0)
	FD_SET(perr[0], &set);
      r = select(max, &set, 0, 0, 0);
      while (r < 0 && errno == EINTR) {
	FD_ZERO(&set);
	if (pout[0] >= 0)
	  FD_SET(pout[0], &set);
	if (perr[0] >= 0)
	  FD_SET(perr[0], &set);
	r = select(max, &set, 0, 0, 0);
      }				/*while */

      if (r < 0) {
	/* assert(errno != EINTR) */
	error("select: %s", strerror(errno));
	exit(1);
      }
      else if (r > 0) {
	if (pout[0] >= 0 && FD_ISSET(pout[0], &set)) {
	  c = read(pout[0], buf, sizeof(buf));
	  if (c > 0) {
	    if (!printflag) {
	      printf("%s:\n", progname);
	      fflush(stdout);
	      printflag = 1;
	    }
	    write(STDOUT_FILENO, buf, c);
	  }
	  else if (c == 0) {
	    close(pout[0]);
	    pout[0] = -1;
	  }
	}
	if (perr[0] >= 0 && FD_ISSET(perr[0], &set)) {
	  c = read(perr[0], buf, sizeof(buf));
	  if (c > 0) {
	    if (!printflag) {
	      fprintf(stderr, "%s:\n", progname);
	      fflush(stderr);
	      printflag = 1;
	    }
	    write(STDERR_FILENO, buf, c);
	  }
	  else if (c == 0) {
	    close(perr[0]);
	    perr[0] = -1;
	  }
	}
      }
      else {
	/* assert(FALSE): select was called with infinite timeout, so
	   it either returns successfully or is interrupted */
      }				/*if */
    }				/*while */
  }

  waitpid(pid, &result, 0);

  if (WIFEXITED(result) && WEXITSTATUS(result)) {
    error("%s exited with return code %d", progname, WEXITSTATUS(result));
    exitstatus = 1;
  }
  else if (WIFSIGNALED(result)) {
    error("%s exited because of uncaught signal %d", progname,
	  WTERMSIG(result));
    exitstatus = 1;
  }
}


/* Find the parts to run & call run_part() */
void run_parts(char *dirname)
{
  struct dirent **namelist;
  char *filename;
  size_t filename_length, dirname_length;
  int entries, i, result;
  struct stat st;

  /* dirname + "/" */
  dirname_length = strlen(dirname) + 1;
  /* dirname + "/" + ".." + "\0" (This will save one realloc.) */
  filename_length = dirname_length + 2 + 1;
  if (!(filename = malloc(filename_length))) {
    error("failed to allocate memory for path: %s", strerror(errno));
    exit(1);
  }
  strcpy(filename, dirname);
  strcat(filename, "/");

  /* scandir() isn't POSIX, but it makes things easy. */
  entries = scandir(dirname, &namelist, valid_name, alphasort);
  if (entries < 0) {
    error("failed to open directory %s: %s", dirname, strerror(errno));
    exit(1);
  }

  i = reverse_mode ? 0 : entries;
  for (i = reverse_mode ? (entries - 1) : 0;
       reverse_mode ? (i >= 0) : (i < entries); reverse_mode ? i-- : i++) {
    if (filename_length < dirname_length + strlen(namelist[i]->d_name) + 1) {
      filename_length = dirname_length + strlen(namelist[i]->d_name) + 1;
      if (!(filename = realloc(filename, filename_length))) {
	error("failed to reallocate memory for path: %s", strerror(errno));
	exit(1);
      }
    }
    strcpy(filename + dirname_length, namelist[i]->d_name);

    strcpy(filename, dirname);
    strcat(filename, "/");
    strcat(filename, namelist[i]->d_name);

    result = stat(filename, &st);
    if (result < 0) {
      error("failed to stat component %s: %s", filename, strerror(errno));
      exit(1);
    }

    if (S_ISREG(st.st_mode)) {
      if (!access(filename, X_OK)) {
	if (test_mode) {
	  printf("%s\n", filename);
	}
	else if (list_mode) {
	  if (!access(filename, R_OK))
	    printf("%s\n", filename);
	}
	else {
	  if (verbose_mode)
	    fprintf(stderr, "run-parts: executing %s\n", filename);
	  run_part(filename);
	  if (exitstatus != 0 && exit_on_error_mode) return;
	}
      }
      else if (!access(filename, R_OK)) {
	if (list_mode) {
	  printf("%s\n", filename);
	}
      }
    }
    else if (!S_ISDIR(st.st_mode)) {
      if (!list_mode) {
	printf("run-parts: component %s is not an executable plain file\n",
	       filename);
	exitstatus = 1;
      }
    }

    free(namelist[i]);
  }
  free(namelist);
  free(filename);
}

/* Process options */
int main(int argc, char *argv[])
{
  umask(022);
  add_argument(0);

  for (;;) {
    int c;
    int option_index = 0;

    static struct option long_options[] = {
      {"test", 0, &test_mode, 1},
      {"list", 0, &list_mode, 1},
      {"verbose", 0, &verbose_mode, 1},
      {"report", 0, &report_mode, 1},
      {"reverse", 0, &reverse_mode, 1},
      {"umask", 1, 0, 'u'},
      {"arg", 1, 0, 'a'},
      {"help", 0, 0, 'h'},
      {"version", 0, 0, 'v'},
      {"lsbsysinit", 0, &lsbsysinit_mode, 1},
      {"exit-on-error", 0, &exit_on_error_mode, 1},
      {0, 0, 0, 0}
    };

    c = getopt_long(argc, argv, "u:ha:v", long_options, &option_index);
    if (c == EOF)
      break;
    switch (c) {
    case 0:
      break;
    case 'u':
      set_umask();
      break;
    case 'a':
      add_argument(optarg);
      break;
    case 'h':
      usage();
      break;
    case 'v':
      version();
      break;
    default:
      fprintf(stderr, "Try `run-parts --help' for more information.\n");
      exit(1);
    }
  }

  /* We require exactly one argument: the directory name */
  if (optind != (argc - 1)) {
    error("missing operand");
    fprintf(stderr, "Try run-parts --help' for more information.\n");
    exit(1);
  }

  if (list_mode && test_mode) {
    error("--list and --test can not be used together");
    fprintf(stderr, "Try run-parts --help' for more information.\n");
    exit(1);
  }

  run_parts(argv[optind]);

  return exitstatus;
}
